extensions-core/lookups-cached-single/src/main/java/io/druid/server/lookup/LoadingLookupFactory.java

```diff
@@ -0,0 +1,145 @@
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.lookup;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.metamx.common.logger.Logger;
+import io.druid.query.lookup.LookupExtractorFactory;
+import io.druid.query.lookup.LookupIntrospectHandler;
+import io.druid.server.lookup.cache.loading.LoadingCache;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@JsonTypeName("loadingLookup")
+public class LoadingLookupFactory implements LookupExtractorFactory
+{
+  private final static Logger LOGGER = new Logger(LoadingLookupFactory.class);
+
+  @JsonProperty("dataFetcher")
+  private final DataFetcher<String, String> dataFetcher;
+
+  @JsonProperty("loadingCacheSpec")
+  private final LoadingCache<String, String> loadingCache;
+
+  @JsonProperty("reverseLoadingCacheSpec")
+  private final LoadingCache<String, List<String>> reverseLoadingCache;
+
+  private final String id =  Integer.toHexString(System.identityHashCode(this));
+  private final LoadingLookup loadingLookup;
+  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  public LoadingLookupFactory(
+      @JsonProperty("dataFetcher") DataFetcher dataFetcher,
+      @JsonProperty("loadingCacheSpec") LoadingCache<String, String> loadingCache,
+      @JsonProperty("reverseLoadingCacheSpec") LoadingCache<String, List<String>> reverseLoadingCache
+  )
+  {
+    this.dataFetcher = Preconditions.checkNotNull(dataFetcher);
+    this.loadingCache = Preconditions.checkNotNull(loadingCache);
+    this.reverseLoadingCache = Preconditions.checkNotNull(reverseLoadingCache);
+    this.loadingLookup = new LoadingLookup(this.dataFetcher, this.loadingCache, this.reverseLoadingCache);
+  }
+
+  @VisibleForTesting
+  protected LoadingLookupFactory(
+      DataFetcher dataFetcher,
+      LoadingCache<String, String> loadingCache,
+      LoadingCache<String, List<String>> reverseLoadingCache,
+      LoadingLookup loadingLookup
+  )
+  {
+    this.dataFetcher = dataFetcher;
+    this.loadingCache = loadingCache;
+    this.reverseLoadingCache = reverseLoadingCache;
+    this.loadingLookup = loadingLookup;
+  }
+
+  @Override
+  public synchronized boolean start()
+  {
+    if (!started.get()) {
+      started.set(loadingLookup.isOpen());
+      LOGGER.info("starting loading lookup [%s]", id);
```

Comment:
Or rather, this is a bit confusing to me, because it is not clear that this log accurately conveys what is happening.